pkg/sdn/plugin/dns_test.go

```diff
@@ -0,0 +1,125 @@
+package plugin
+
+import (
+	"testing"
+	"time"
+)
+
+func TestAddDNS(t *testing.T) {
+	type dnsTest struct {
+		testCase string
+
+		domainName    string
+		dnsValueSet   bool
+		expectFailure bool
+	}
+
+	tests := []dnsTest{
+		{
+			testCase:      "Test valid domain name",
+			domainName:    "example.com",
+			dnsValueSet:   true,
+			expectFailure: false,
+		},
+		{
+			testCase:      "Test invalid domain name",
+			domainName:    "sads@#$.com",
+			dnsValueSet:   false,
+			expectFailure: true,
+		},
+	}
+
+	for _, test := range tests {
+		dns := DNS{}
+		err := dns.Add(test.domainName)
+		if test.expectFailure && err == nil {
+			t.Fatalf("Test case: %s failed, expected failure but got success", test.testCase)
+		} else if !test.expectFailure && err != nil {
+			t.Fatalf("Test case: %s failed, err: %v", test.testCase, err)
+		}
+
+		if test.dnsValueSet {
+			d, ok := dns.dnsMap[test.domainName]
+			if !ok {
+				t.Fatalf("Test case: %s failed, domain %q not found in dns map", test.testCase, test.domainName)
+				continue
+			}
+			if len(d.ips) == 0 || d.ttl.Seconds() == 0 && d.nextQueryTime.Second() == 0 {
+				t.Fatalf("Test case: %s failed, ips/ttl/nextQueryTime for the domain %q is not set", test.testCase, test.domainName)
+			}
+		} else {
+			if _, ok := dns.dnsMap[test.domainName]; ok {
+				t.Fatalf("Test case: %s failed, unexpected domain %q found in dns map", test.testCase, test.domainName)
+			}
+		}
+	}
+}
+
+func TestUpdateDNS(t *testing.T) {
+	type dnsTest struct {
+		testCase string
+
+		domainName string
+		dnsMapSize int
+
+		// Only valid for dnsMapSize > 0
+		expectIPChange            bool
+		expectTTLChange           bool
+		expectNextQueryTimeChange bool
+		expectFailure             bool
+	}
+
+	tests := []dnsTest{
+		{
+			testCase:                  "Test dns update of valid domain",
+			domainName:                "example.com",
+			dnsMapSize:                1,
```

Comment:
Assuming that example.com will always only return 1 IP isn't safe.

I think a better way to do this would be to use kexec rather than os/exec in dns.go so you can test it against pre-recorded dig output. (eg, see pkg/util/ipcmd/ipcmd_test.go).

That can happen in a later commit (so this PR doesn't get delayed any more), but add a github issue or trello card to track it. Not sure if I think it would be better to commit this test as-is for now, or to just remove it from the PR for now.